### PR TITLE
[JENKINS-25735, JENKINS-41955, etc.] Improve masking of passwords in the plugin

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] Rework the broken implementation of the `ParameterValue` classes aching in `MaskPasswordConfig`. The implementation was unreliable since Mask Password `2.5`
- [x] Rework the mechanism of guessing sensitive `ParameterValue`s by classnames of `ParameterDefinition`s. This is still an architecture bug IMHO, but now the engine guesses classes more reliably using all available `ParameterDefinition` method overrides
- [x] [JENKINS-41955](https://issues.jenkins-ci.org/browse/JENKINS-41955) - Replace the hotfix made by @bpmarinho in https://github.com/jenkinsci/mask-passwords-plugin/pull/11 . The new parameter value implementation guesses the case reliably
- [x] [JENKINS-25735](https://issues.jenkins-ci.org/browse/JENKINS-25735) - Always consider sensitive parameter values as values, which require masking
- [x] Tests, Tests, Tests

Likely the change also solves many other issues reported to the plugin. Still need direct tests for them BTW.

@reviewbybees @bpmarinho @brockoffdev I would appreciate reviews and manual testing

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@reviewbybees done (review timeout)

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers I will appreciate feedback

File: src/test/java/com/michelin/cio/hudson/plugins/integrations/CorePasswordParameterTest.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        MaskPasswordsBuildWrapper maskPasswordsBuildWrapper
+                = new MaskPasswordsBuildWrapper(Collections.<MaskPasswordsBuildWrapper.VarPasswordPair>emptyList());
+        project.getBuildWrappersList().add(maskPasswordsBuildWrapper);
+
+        project.getBuildersList().add(new TestBuilder() {
+            @Override
+            public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+                listener.getLogger().println(logWithClearTextPassword);
+                build.setResult(Result.SUCCESS);
```
Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/integrations/CorePasswordParameterTest.java]: gratuitous

File: src/test/java/com/michelin/cio/hudson/plugins/passwordparam/PasswordParameterTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        final String logWithClearTextPassword = "printed " + clearTextPassword + " oops";
+        final String logWithHiddenPassword = "printed ******** oops";
+
+        FreeStyleProject project
+                = j.jenkins.createProject(FreeStyleProject.class, "testPasswordParameter");
+
+        PasswordParameterDefinition passwordParameterDefinition = new PasswordParameterDefinition("Password1", null);
+        ParametersDefinitionProperty parametersDefinitionProperty
+                = new ParametersDefinitionProperty(passwordParameterDefinition);
+        project.addProperty(parametersDefinitionProperty);
```
Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/passwordparam/PasswordParameterTest.java]: Why the copy-pasted code? Better to factor out the common parts.

**Comment by oleg-nenashev** [comment]
@reviewbybees done, merging as a maintainer

**Comment by bpmarinho** [comment]
Just tested with 2.10.1 and it's working as expected regarding the fix I did earlier and was redesigned with this change.
Use **jenkins:2.19.4-alpine** and **jenkins:1.609.3** for manual testing.
Thanks for getting out a new release with the fix.

**Comment by oleg-nenashev** [comment]
Thanks @bpmarinho ! I had to tweak the implementation "a bit" to work it more reliably though there are still some issues with automatic Secret discovery IMHO

**Pull request merged**